### ISSUE TITLE
Revisit server1/server2 and case1/case2 and value1/value2 naming in Servers, oneOf, allOf/anyOf

### ISSUE THREAD

**Issue by czechboy0** [apple/swift-openapi-generator#28]
Revisit server1/server2 and case1/case2 and value1/value2 naming in Servers, oneOf, allOf/anyOf.

We don't like the names, we should derive better unique names, e.g. based on:
- Swift type
- JSON path
- user-provided description
- something else inherent

We just have to keep in mind that in many cases (e.g. inline schemas), we don't have a unique name to use, and might need to derive one - there we need to watch out for potential naming conflicts.

**Comment by czechboy0** [comment]
@simonjbeaumont to gather some data.

**Comment by simonjbeaumont** [comment]
# Precedence for geneating types using schema composition

## Background

- https://spec.openapis.org/oas/v3.0.3#schema-object
- https://json-schema.org/understanding-json-schema/reference/combining.html#anyof

## Inputs

I decided to use the following snippet of OpenAPI YAML for this research:

```yaml
components:
  schemas:
    Badger:
      type: object
      properties:
        b:
          type: boolean
    Mushroom:
      type: object
      properties:
        m:
          type: integer
    Snake:
      type: object
      properties:
        s:
          type: string
    MyAnyOf:
      anyOf:
      - $ref: '#/components/schemas/Badger'
      - $ref: '#/components/schemas/Mushroom'
      - $ref: '#/components/schemas/Snake'
    MyAllOf:
      allOf:
      - $ref: '#/components/schemas/Badger'
      - $ref: '#/components/schemas/Mushroom'
      - $ref: '#/components/schemas/Snake'
    MyOneOf:
      oneOf:
      - $ref: '#/components/schemas/Badger'
      - $ref: '#/components/schemas/Mushroom'
      - $ref: '#/components/schemas/Snake'
```

## Generators

I'll use the following set of generators initially:

- `openapi-generator -g swift`
- `openapi-generator -g go`
- `openapi-generator -g rust`
- `openapi-typescript`
- `openapi-typescript-codegen`

## Outputs

I've tried to capture just enough of the output for each generator to draw themes about the question at hand.

### `openapi-generator -g swift`

```swift
struct MyAnyOf: Codable, JSONEncodable, Hashable {
    var b: Bool?
    var m: Int?
    var s: String?
}

struct MyAllOf: Codable, JSONEncodable, Hashable {
    var b: Bool?
    var m: Int?
    var s: String?
}

public enum MyOneOf: Codable, JSONEncodable, Hashable {
    case typeBadger(Badger)
    case typeMushroom(Mushroom)
    case typeSnake(Snake)
}
```

### `openapi-generator -g go`

```go
type MyAnyOf struct {
	Badger *Badger
	Mushroom *Mushroom
	Snake *Snake
}

type MyAllOf struct {
	B *bool `json:"b,omitempty"`
	M *int32 `json:"m,omitempty"`
	S *string `json:"s,omitempty"`
}

type MyOneOf struct {
	Badger *Badger
	Mushroom *Mushroom
	Snake *Snake
}
```

### `openapi-generator -g rust`

```rust
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct MyAnyOf {
    #[serde(rename = "b", skip_serializing_if = "Option::is_none")]
    pub b: Option<bool>,
    #[serde(rename = "m", skip_serializing_if = "Option::is_none")]
    pub m: Option<i32>,
    #[serde(rename = "s", skip_serializing_if = "Option::is_none")]
    pub s: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct MyAllOf {
    #[serde(rename = "b", skip_serializing_if = "Option::is_none")]
    pub b: Option<bool>,
    #[serde(rename = "m", skip_serializing_if = "Option::is_none")]
    pub m: Option<i32>,
    #[serde(rename = "s", skip_serializing_if = "Option::is_none")]
    pub s: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct MyOneOf {
    #[serde(rename = "b", skip_serializing_if = "Option::is_none")]
    pub b: Option<bool>,
    #[serde(rename = "m", skip_serializing_if = "Option::is_none")]
    pub m: Option<i32>,
    #[serde(rename = "s", skip_serializing_if = "Option::is_none")]
    pub s: Option<String>,
}
```

### `openapi-typescript`

```typescript
export interface components {
  schemas: {
    Badger: {
      b?: boolean;
    };
    Mushroom: {
      m?: number;
    };
    Snake: {
      s?: string;
    };
    MyAnyOf: components["schemas"]["Badger"] | components["schemas"]["Mushroom"] | components["schemas"]["Snake"];
    MyAllOf: components["schemas"]["Badger"] & components["schemas"]["Mushroom"] & components["schemas"]["Snake"];
    MyOneOf: components["schemas"]["Badger"] | components["schemas"]["Mushroom"] | components["schemas"]["Snake"];
  };
  responses: never;
  parameters: never;
  requestBodies: never;
  headers: never;
  pathItems: never;
}
```


### `openapi-typescript-codegen`

```typescript
export type MyAnyOf = (Badger | Mushroom | Snake);
export type MyAllOf = (Badger & Mushroom & Snake);
export type MyOneOf = (Badger | Mushroom | Snake);
```

## Learnings

The above outputs show different strategies for handling schema composition,
including flattened types, union types, variant types, etc.

However, none of them seem to need to make use of an index-based naming, e.g.
`case1`, `case2`, etc.

**Comment by czechboy0** [comment]
Yeah if constrained to just references to objects as subschemas, those can be handled pretty well, because they have a global name.

But all of these aggregate types (except for oneOfs with a discriminator) can have any schema as a subschemas, and those types can be inlined (in which case you don't have a global name) and don't even have to be unique (you can have two string schemas that only differ by their regex constraint).

Curious to see the output for the following additional subschemas:

```yaml
- type: string
- type: string
- type: string
  enum:
    - foo
- type: integer
- type: object
- type: object
  properties:
    foo:
      type: string
```

I also suspect that variadic generics in Swift could help us here, but all this was designed before they were available.

**Comment by simonjbeaumont** [comment]
Those set of subschemas won't work for all of the schema composition methods, because they cannot be used in conjunction in some cases in an OAS-compliant way. However, I threw them into the generators and updated the comment with the findings.

Still no index-based typenames.

**Comment by czechboy0** [comment]
Sure, in the case of allOf and anyOf, there are some mutually incompatible schemas, but the oneOf should be able to support all 10 schemas. From the above, it seems none of the generators were able to generate more than 8, and some stopped at 7.

And getting all the way to 10 is where they'd have to choose either an order-based name, or create a unique name from the contents, such as `string_with_pattern_backslash_d_minItems_10_...` etc. The order-based name is a lot simpler to implement, but the content-based name would also work.

Would love to see a generator that can actually handle all 10 cases, and which method of naming they chose.

**Comment by simonjbeaumont** [comment]
> but the oneOf should be able to support all 10 schemas.

I don't think that's true. For example you suggested these:

```yaml
- type: string
- type: string
  pattern: '\d'
- type: string
  enum:
    - foo
```

which cannot work because it needs to satisfy (exactly one)[^1]. But the first is always satisfied, which is why I think some of the generators skipped it (there was some diagnostic output to this effect too).

[^1]: https://json-schema.org/understanding-json-schema/reference/combining.html#id7

**Comment by czechboy0** [comment]
Good catch, let me amend to:

```yaml
- type: string
   pattern: '\w'
- type: string
  pattern: '\d'
- type: string
  enum:
    - foo
```

**Comment by jordanebelanger** [comment]
I ran into a similar case today and I would suggest using the json-schema compliant annotation "title"  https://json-schema.org/understanding-json-schema/reference/annotations

In case of collision, simply default to `case1` like we do today, perhaps with a warning 🤷

**Comment by czechboy0** [comment]
Hi @jordanebelanger,

> In case of collision

so far we've avoided naming entities based on their sibling entities. I.e. a name of an entity depending on a naming conflict. The reason for that is because then renaming one entity _changes_ the name of another entity as well, likely leading to an unexpected API break.

Our naming logic, for now, has these requirements:
- purely based on the named entity (and optionally its parents), but not siblings (i.e. no changing of names in case of conflicts, the naming scheme itself must avoid conflicts)
- deterministic, easy to guess by just reading the OpenAPI doc

The issue with `title` is that it's not required, and it doesn't have to be unique. Those two properties don't make it a very good signal for naming.

That said - the `title` should appear at least in the generated doc comment - are you seeing it there?

**Comment by nstringham** [comment]
Open API 3.2 adds a `name` to the server that could be good for this.
https://learn.openapis.org/specification/servers.html#enhanced-server-fields
